docs/source/ckpt_converters/dev_guide.rst

```diff
@@ -0,0 +1,234 @@
+
+     .. code-block:: text
+
+        Map the following key names and tensor shapes from Model A to their equivalents in Model B. Here is an example mapping: Model A's 'model.layer.weight' corresponds to Model B's 'module.block.weight'.
+        ============================================================
+        embedder.weight torch.Size([256128, 2048])
+        ...
+        ============================================================
+
+   Based on the results, update the following code accordingly:
+
+   .. code-block:: python
+
+       def create_rename_keys(num_hidden_layers):
+           rename_keys = []
+           for i in range(num_hidden_layers):
+               # encoder layers: output projection, 2 feedforward neural networks, and 2 layernorms
+               # @chatgpt to fill in layer-dependent keys above
+
+           # @chatgpt fill in non-layer-dependent keys above
+           rename_keys.extend(
+               [
+                   # ...
+               ]
+           )
+
+           return rename_keys
+
+   **Note**: Also list all the keys not included in the conversion above.
+
+2. **Common issues when converting: results not matching between Community model and NeMo model**:
+
+   a. Megatron Core uses a special QKV layout, which needs careful handling and reshaping from community models, especially when GQA or MQA is used. Refer to the `Gemma Huggingface to NeMo converter <https://github.com/NVIDIA/NeMo/tree/main/scripts/checkpoint_converters/convert_gemma_hf_to_nemo.py#L144>`_ for guidance.
+
+   b. GLU Variants weights could also be a common source of error. In Megatron Core, the regular feedforward projection weights and gated forward weights are fused together, requiring careful attention to the order of these two. Refer to the `Gemma Huggingface to NeMo converter <https://github.com/NVIDIA/NeMo/tree/main/scripts/checkpoint_converters/convert_gemma_hf_to_nemo.py#L135>`_ for more details.
+
+3. The ``create_hf_model`` function can be used to create a model programmatically. For reproducibility, see the example provided at `GitHub <https://github.com/NVIDIA/NeMo/blob/main/tests/setup/models/create_hf_model.py>`_. This function creates a randomly initialized HuggingFace model for testing purposes. The model can be specified by name or path for creating its config and tokenizer using HuggingFace transformers AutoConfig and AutoTokenizer functions.
+
+Example usage:
+
+.. code-block:: python
+
+    create_hf_model(
+        model_name_or_path="/home/TestData/nlp/meta-llama/Llama-2-7b-hf",
+        output_dir=os.path.join(args.save_dir, "megatron_llama/llama-ci-hf"),
+        config_updates={
+            "hidden_size": 256,
+            "num_attention_heads": 4,
+            "num_hidden_layers": 2,
+            "num_key_value_heads": 4
+        },
+        overwrite=args.overwrite,
```

Comment:
* we could remove `args.overwrite` to make this snippet functional (and stick to default value)
* possibly also use` model_name_or_path="meta-llama/Llama-2-7b-hf"` to refer to the original HF model directly instead of some local path